src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -1349,6 +1142,11 @@ private void RerouteAllQueuedMessages(ActivationData activation, ActivationAddre
             {
                 RequestContext.Import(requestContextData);
                 await activation.Lifecycle.OnStart();
+                // TODO: Consider pre and post activation stages in lifecycle for PreActivateProtocolParticipant and PostActivateProtocolParticipant like behaviors
```

Comment:
Yes, I was even considering that the `Grain.OnActivateAsync` could actually be implicitly hooked up by the `Grain` base class being a lifecycle participant, not as a special method that we handle separately in the catalog.
This way, we get closer to the goal of supporting POCO objects as grains in the future, if all that is in `Grain` can be implemented by scoped activation services and lifetime observers.

Reply:
That is the plan.  All of grain lifecycle should be handled within the lifecycle framework.  We've just not refactored it all yet.  A good next step would be to add pre-activate, activate, and post-activate stages to the lifecycle, and move the onactivate/deavtivate within the lifecycle (as you've describe) along with the pre/post activate logic.  Though, I'm not convinced we need the pre/post activate, so I'd like to understand why that was needed.